### PR TITLE
New interface about conn_param structure.

### PR DESCRIPTION
**Pull request by wanghaEMQ**

### CODE REVIEW

**Comment by wanghaEMQ** [comment]
Add refcnt to conn_param structure.

File: include/nng/protocol/mqtt/mqtt_parser.h [hunk 1]
```diff
@@ -37,10 +37,10 @@ extern void       pub_extra_set_packet_id(pub_extra *, uint16_t);
...
 // MQTT CONNECT
 int32_t conn_handler(uint8_t *packet, conn_param *conn_param);
-int     new_conn_param(conn_param **cparam);
-void    init_conn_param(conn_param *cparam);
-void    destroy_conn_param(conn_param *cparam);
-void    deep_copy_conn_param(conn_param *new_cp, conn_param *cp);
+int     conn_param_alloc(conn_param **cparam);
+void    conn_param_free(conn_param *cparam);
+void    conn_param_clone(conn_param *cparam);
+// void    deep_copy_conn_param(conn_param *new_cp, conn_param *cp);
```
Comment (JaylinYu) [inline include/nng/protocol/mqtt/mqtt_parser.h:43]: just delete it

Comment (JaylinYu) [inline include/nng/protocol/mqtt/mqtt_parser.h:43]: just delete it

File: src/sp/protocol/mqtt/mqtt_parser.c [hunk 2]
```diff
@@ -683,23 +681,64 @@ conn_handler(uint8_t *packet, conn_param *cparam)
...
 
 void
-destroy_conn_param(conn_param *cparam)
+conn_param_free(conn_param *cparam)
 {
 	if (cparam == NULL) {
 		return;
 	}
+	if (nni_atomic_dec_nv(&cparam->refcnt) != 0) {
+		return;
```
Comment (JaylinYu) [inline src/sp/protocol/mqtt/mqtt_parser.c:740]: what if double free?

File: src/sp/transport/tcp/tcp.c [hunk 3]
```diff
@@ -334,7 +334,7 @@ tcptran_pipe_nego_cb(void *arg)
 		} else {
 			rv = NNG_EPROTO;
 			nng_free(p->conn_buf, p->wantrxhead);
-			nng_free(p->tcp_cparam, sizeof(struct conn_param));
+			conn_param_free(p->tcp_cparam);
```
Comment (JaylinYu) [inline src/sp/transport/tcp/tcp.c:337]: it must be used with mutex then

**Review by JaylinYu — APPROVED** [review]

File: src/sp/protocol/mqtt/mqtt_parser.c (hunk 2 above)
Comment (wanghaEMQ) [reply src/sp/protocol/mqtt/mqtt_parser.c:740]: Program carefully.

**Pull request merged**